### PR TITLE
Allow proxy

### PR DESCRIPTION
**Pull request by andresespinosapc**
I add a proxyOptions optional parameter to use a SOCKS5 proxy. I haven't assured that it's really sending all through the proxy, but I guess it works because if I put incorrect credentials it throws the expected exception.

### CODE REVIEW

File: src/transport/transport.ts [hunk 1]
```diff
@@ -1,9 +1,11 @@
+import { SocksProxy } from 'socks';
+
 export abstract class Transport<T> {
     /**
      * This will be set by the MqttClient
      */
     public callbacks: TransportConnectOptions;
-    public constructor(protected options: T) {}
+    public constructor(protected options: T, protected proxyOptions?: TransportProxyOptions) {}
```
Comment (Nerixyz) [inline src/transport/transport.ts]: The `proxyOptions` should be in the `options` object for a given `Transport` as not every transport supports a proxy.

File: src/transport/tls.transport.ts [hunk 2]
```diff
@@ -1,18 +1,34 @@
 import { Transport } from './transport';
 import { TLSSocket, connect } from 'tls';
 import * as URL from 'url';
+import { SocksClient } from 'socks';
 
 export class TlsTransport extends Transport<{ url: string; enableTrace?: boolean }> {
```
Comment (Nerixyz) [inline src/transport/tls.transport.ts]: The `proxyOptions` should be here.

**Review by Nerixyz — CHANGES_REQUESTED** [review]

File: src/transport/tls.transport.ts (hunk 2 above)
Comment (andresespinosapc) [reply src/transport/tls.transport.ts]: Ok! Didn't see that. And where do you think I should put the TransportProxyOptions type? Should I rename it also?

Comment (Nerixyz) [reply src/transport/tls.transport.ts]: > And where do you think I should put the TransportProxyOptions type? Should I rename it also?

These should be in the `tsl.transport.ts` file. You may rename it to `TlsTransportProxyOptions`.

**Review by Nerixyz — APPROVED** [review]

**Pull request merged**